### PR TITLE
Add '-concurrency' flag and logic to limit concurrent scrape

### PR DESCRIPTION
**Pull request by obaudys**
This change adds a flag to limit the concurrent HTTP requests the exporter makes when scraping per-database and per-view stats.  This might be useful for people with thousand of DBs or views.

I've tested quite extensively on my setup, `-concurrency=1` takes me 17s to scrape; whereas 5 takes me 7s, etc. up to 3.5s with 0 (unlimited)

### CODE REVIEW

**Comment by gesellix** [comment]
That last refactoring made your PR much more beautiful, thanks!

I'll make another and more detailed review soonish, I want to share some quick thoughts here. Please consider them as _thoughts_ and not necessarily as change requests. Some aspects can be solved afterwards.

- The logic becomes more advanced now, so some high level tests might make sense.
- Having a quick look at https://medium.com/@zufolo/a-pattern-for-limiting-the-number-of-goroutines-in-execution-56e13b226e72 I'm thinking about edge cases like `numberOfWorkers < numberOfJobs`, which might lead to deadlocks. This is something tests might validate and I'll have to keep that in mind when making the actual review.
- Maybe some of the goroutines can be pulled up and get proper function names so that the code readability improves. Can be refactored independent to this PR.

File: couchdb-exporter.go [hunk 1]
```diff
@@ -124,6 +125,13 @@ func init() {
 			Hidden:      false,
 			Destination: &exporterConfig.databaseViews,
 		}),
+		altsrc.NewUintFlag(cli.UintFlag{
+			Name:        "concurrency",
```
Comment (gesellix) [inline couchdb-exporter.go]: We need to discuss this name. I'd prefer a more specific name, something along the idea of `concurrent-database-requests`.

Comment (gesellix) [reply couchdb-exporter.go]: ... or even `database-concurrent-requests`, to keep the name similar to the other parameters.

Comment (obaudys) [reply couchdb-exporter.go]: Sure, I'll make a commit when I get time.

**Comment by obaudys** [comment]
@gesellix Thanks for reviewing the change.  Regarding your comments:

- I'll have a crack at making some tests in a subsequent commit.
- In this case the `numberOfWorkers == numberOfJobs`,  the workers just wait on a semaphore before proceeding, or abort if the parent goroutine has indicated it received an error from another worker with `semaphore.Abort()`.
- I'll think about how I might turn the closures into proper functions.  Right now they capture some variables from the enclosing scope; namely the semaphore, the channel for the results / errors, and the name of the db/view to query;  these can be passed in as parameters I guess.

**Comment by obaudys** [comment]
@gesellix I've added some tests to confirm the semaphore works as intended;   I'm not sure that's what you meant by high level tests, but I had a lot of fun writing these. 

Did you mean to test the actual database and view scrapes?  When I was testing for goroutine leaks in the face of error,  I injected errors by adding
`if rand.Intn(200) == 57 {
    err = fmt.Errorf("Random madeup error")
}`
after every `data, err := c.Request(...)` to simulate occasional errors, then I ran the exporter for a while with `GODEBUG=gctrace=1`, watching for increasing heap over time.  I'm not sure how to test for that in an actual unit test.

**Comment by obaudys** [comment]
@gesellix any chance you will merge this pull request?  I have more contributions I want to make,  but they kind of depend on this branch.  Should I just commit them as part of this pull request, or wait till this is accepted / rejected?

**Comment by gesellix** [comment]
@obaudys sorry for the delay! As far as I followed your recent changes I consider merging your pull request this weekend. I'm going to make a more detailed review, though.

**Comment by gesellix** [comment]
> ... I'm not sure how to test for that in an actual unit test.

Same here, I also still haven't found a good way to perform such an automated test, so I would try to find a solution afterwards. I appreciate your manual checks!

File: couchdb-exporter.go [hunk 2]
```diff
@@ -124,6 +125,13 @@ func init() {
 			Hidden:      false,
 			Destination: &exporterConfig.databaseViews,
 		}),
+		altsrc.NewUintFlag(cli.UintFlag{
+			Name:        "database.concurrent.requests",
+			Usage:       "maximum concurrent calls to couchDB, or 0 for unlimited",
```
Comment (gesellix) [inline couchdb-exporter.go]:
```suggestion
			Usage:       "maximum concurrent calls to CouchDB, or 0 for unlimited",
```

File: lib/couchdb-client.go [hunk 3]
```diff
@@ -327,45 +345,65 @@ func (c *CouchdbClient) enhanceWithViewUpdateSeq(dbStatsByDbName map[string]Data
...
-							"limit=0",
-						}, "&")
-						var viewDoc ViewResponse
-						viewDocData, err := c.Request("GET", fmt.Sprintf("%s/%s/%s/_view/%s?%s", c.BaseUri, dbName, row.Doc.Id, viewName, query), nil)
-						err = json.Unmarshal(viewDocData, &viewDoc)
-						if err != nil {
-							v <- viewresult{err: fmt.Errorf("error unmarshalling view doc for view '%s/%s/_view/%s': %v", dbName, row.Doc.Id, viewName, err)}
+				row := row
+				go func() {
+					defer func() { done <- struct{}{} }()
```
Comment (gesellix) [inline lib/couchdb-client.go]:
```suggestion
					defer func() {
						done <- struct{}{}
					}()
```

File: lib/couchdb-client.go [hunk 4]
```diff
@@ -327,45 +345,65 @@ func (c *CouchdbClient) enhanceWithViewUpdateSeq(dbStatsByDbName map[string]Data
...
+					}
+					v := make(chan viewresult, len(row.Doc.Views))
+					for viewName := range row.Doc.Views {
+						viewName := viewName
+						go func() {
+							//klog.Infof("/%s/%s/_view/%s\n", dbName, row.Doc.Id, viewName)
+							err := semaphore.Acquire()
+							if err != nil {
+								// send something to parent coroutine so it doesn't block forever on receive
+								v <- viewresult{err: fmt.Errorf("Aborted view stats for /%s/%s/_view/%s", dbName, row.Doc.Id, viewName)}
```
Comment (gesellix) [inline lib/couchdb-client.go]:
```suggestion
								v <- viewresult{err: fmt.Errorf("aborted view stats for /%s/%s/_view/%s", dbName, row.Doc.Id, viewName)}
```

File: lib/couchdb-client.go [hunk 5]
```diff
@@ -511,3 +550,58 @@ func NewCouchdbClient(uri string, basicAuth BasicAuth, insecure bool) *CouchdbCl
...
+		s.sem <- struct{}{}
+	}
+	return s
+}
+
+// Acquire the semaphore; blocks until ready, or returns error to indicate the goroutine should abort
+func (s Semaphore) Acquire() error {
+	select {
+	case <-s.abort:
+		return fmt.Errorf("Could not acquire semaphore")
```
Comment (gesellix) [inline lib/couchdb-client.go]:
```suggestion
		return fmt.Errorf("could not acquire semaphore")
```

File: lib/couchdb-client.go [hunk 6]
```diff
@@ -511,3 +550,58 @@ func NewCouchdbClient(uri string, basicAuth BasicAuth, insecure bool) *CouchdbCl
...
+func (s Semaphore) Release() {
+	if s.concurrency > 0 {
+		select {
+		case s.sem <- struct{}{}:
+		default: // should not happen unless someone double released
+		}
+	}
+}
+
+// Signal abort for anyone waiting on the Semaphor
```
Comment (gesellix) [inline lib/couchdb-client.go]:
```suggestion
// Signal abort for anyone waiting on the Semaphore
```

File: lib/couchdb-client_test.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+	// Test3: test serialised workers, ie. 1 at a time;
+	sem := NewSemaphore(1)
+	results := make(chan struct{}, 100)
+	sem.Acquire()
+	for i := 0; i < 100; i++ {
+		go worker(sem, results)
+	}
+	sem.Release()
+	time.Sleep(20 * time.Millisecond)
+	if len(results) > 20 { // should never complete this many in the alotted time
```
Comment (gesellix) [inline lib/couchdb-client_test.go]:
```suggestion
	if len(results) > 20 { // should never complete this many in the allotted time
```

**Comment by obaudys** [comment]
@gesellix I've committed those suggestions, thanks.

This is not really relevant to this pull requests, but I'd thought I'd ask here anyway since it bothers me: what is the purpose of the following construct?

`*&exporterConfig.databaseViews`

It seems to me that the `*` and the `&` cancel each other out, making them pointless.  I have been unable to find an explanation by googling, it does not seem to be a commonly used idiom.

**Comment by gesellix** [comment]
Regarding the `*&...`: you're right, I also don't know why it is like that. I'm going to remove them. One reason might be that I've written that code in my early beginnings of learning Golang, so, well, let's consider them as legacy ;)

**Comment by obaudys** [comment]
@gesellix you can automate the removal of `*&...` with
`
gofmt -r '*&a -> a' -w couchdb-exporter.go 
`
;-)

**Comment by gesellix** [comment]
Seems like I can learn Golang from you, great tip! :)
